Cubical/Structures/Poset.agda

```diff
@@ -0,0 +1,189 @@
+isTransitive {ℓ₀ = ℓ₀} {ℓ₁ = ℓ₁} {A = X} _⊑_ = φ , φ-prop
+  where
+    φ      : Type (ℓ-max ℓ₀ ℓ₁)
+    φ      = ((x y z : X) → [ x ⊑ y ⇒ y ⊑ z ⇒ x ⊑ z ])
+    φ-prop : isProp φ
+    φ-prop = isPropΠ λ x → isPropΠ λ y → isPropΠ λ z →
+               snd (x ⊑ y ⇒ y ⊑ z ⇒ x ⊑ z)
+
+isAntisym : {A : Type ℓ₀} → isSet A → Order ℓ₁ A → hProp (ℓ-max ℓ₀ ℓ₁)
+isAntisym {ℓ₀ = ℓ₀} {ℓ₁ = ℓ₁} {A = X} A-set _⊑_ = φ , φ-prop
+  where
+    φ      : Type (ℓ-max ℓ₀ ℓ₁)
+    φ      = ((x y : X) → [ x ⊑ y ] → [ y ⊑ x ] → x ≡ y)
+    φ-prop : isProp φ
+    φ-prop = isPropΠ λ x → isPropΠ λ y → isPropΠ λ _ → isPropΠ λ _ → A-set x y
+
+-- The predicate expressing that a given order satisfies the partial order
+-- axioms.
+satPosetAx : (ℓ₁ : Level) (A : Type ℓ₀) → Order ℓ₁ A → hProp (ℓ-max ℓ₀ ℓ₁)
+satPosetAx {ℓ₀ = ℓ₀} ℓ₁ A _⊑_ = φ , φ-prop
+  where
+    isPartial : isSet A → hProp (ℓ-max ℓ₀ ℓ₁)
+    isPartial A-set = isReflexive _⊑_ ⊓ isTransitive _⊑_ ⊓ isAntisym A-set _⊑_
+
+    φ         = Σ[ A-set ∈ isSet A ] [ isPartial A-set ]
+    φ-prop    = isOfHLevelΣ 1 isPropIsSet (λ x → snd (isPartial x))
+
+-- The poset structure.
+PosetStr : (ℓ₁ : Level) → Type ℓ₀ → Type (ℓ-max ℓ₀ (ℓ-suc ℓ₁))
+PosetStr ℓ₁ = add-to-structure (Order ℓ₁) λ A _⊑_ → [ satPosetAx ℓ₁ A _⊑_ ]
+
+PosetStr-set : (ℓ₁ : Level) (A : Type ℓ₀) → isSet (PosetStr ℓ₁ A)
+PosetStr-set ℓ₁ A =
+  isSetΣ
+    (isSetΠ λ _ → isSetΠ λ _ → isSetHProp) λ _⊑_ →
+      isProp→isSet (snd (satPosetAx ℓ₁ A _⊑_))
+
+Poset : (ℓ₀ ℓ₁ : Level) → Type (ℓ-max (ℓ-suc ℓ₀) (ℓ-suc ℓ₁))
+Poset ℓ₀ ℓ₁ = Σ (Type ℓ₀) (PosetStr ℓ₁)
+
+-- Some projections for syntactic convenience.
+
+-- Carrier set of a poset.
+∣_∣ₚ : Poset ℓ₀ ℓ₁ → Type ℓ₀
+∣ X , _ ∣ₚ = X
+
+strₚ : (P : Poset ℓ₀ ℓ₁) → PosetStr ℓ₁ ∣ P ∣ₚ
+strₚ (_ , s) = s
+
+rel : (P : Poset ℓ₀ ℓ₁) → ∣ P ∣ₚ → ∣ P ∣ₚ → hProp ℓ₁
+rel (_ , _⊑_ , _) = _⊑_
+
+syntax rel P x y = x ⊑[ P ] y
+
+⊑[_]-refl : (P : Poset ℓ₀ ℓ₁) → (x : ∣ P ∣ₚ) → [ x ⊑[ P ] x ]
+⊑[_]-refl (_ , _ , _ , ⊑-refl , _) = ⊑-refl
+
+⊑[_]-trans : (P : Poset ℓ₀ ℓ₁) (x y z : ∣ P ∣ₚ)
+           → [ x ⊑[ P ] y ] → [ y ⊑[ P ] z ] → [ x ⊑[ P ] z ]
+⊑[_]-trans (_ , _ , _ , _ , ⊑-trans , _) = ⊑-trans
+
+⊑[_]-antisym : (P : Poset ℓ₀ ℓ₁) (x y : ∣ P ∣ₚ)
+             → [ x ⊑[ P ] y ] → [ y ⊑[ P ] x ] → x ≡ y
+⊑[_]-antisym (_ , _ , _ , _ , _ , ⊑-antisym) = ⊑-antisym
+
+carrier-is-set : (P : Poset ℓ₀ ℓ₁) → isSet ∣ P ∣ₚ
+carrier-is-set (_ , _ , is-set , _) = is-set
+
+module PosetReasoning (P : Poset ℓ₀ ℓ₁) where
+
+  _⊑⟨_⟩_ : (x : ∣ P ∣ₚ) {y z : ∣ P ∣ₚ}
+         → [ x ⊑[ P ] y ] → [ y ⊑[ P ] z ] → [ x ⊑[ P ] z ]
+  _ ⊑⟨ p ⟩ q = ⊑[ P ]-trans _ _ _ p q
+
+  _■ : (x : ∣ P ∣ₚ) → [ x ⊑[ P ] x ]
+  _■ = ⊑[ P ]-refl
+
+  infixr 0 _⊑⟨_⟩_
+  infix  1 _■
+
+-- Poset univalence.
+RP-iso-prop : (P Q : Σ (Type ℓ₀) (Order ℓ₁))
+            → (i : fst P ≃ fst Q) → isProp (order-iso P Q i)
+RP-iso-prop (A , _⊑₀_) (B , _⊑₁_) i =
+    isPropΠ λ x → isPropΠ λ y → snd (x ⊑₀ y ⇔ f x ⊑₁ f y)
+  where
+    f = equivFun i
```

Comment:
Pattern-match on i?

Reply:
I wonder why I was not doing this already. Did the older versions require me to do this for some reason? I think I remember something like that.